tech_notes/tn008_auto_insert_cells.md

```diff
@@ -0,0 +1,395 @@
+generate one or more cells to be inserted after the current cell. This is similar to GitHub Copilot and 
+[Continue.Dev Autocomplete](https://docs.continue.dev/walkthroughs/tab-autocomplete. An important difference
+is that we don't need Foyle to autocomplete the current cell. This should simplify the problem because 
+* we have more time to generate completions
+  * the completion needs to be ready by the time the user is ready to move onto the next cell rather than before
+    they type the next character
+* we can use larger models and longer context windows since we lave a larger latency budget
+* we don't need to check that the completion is still valid after each character is typed into the current cell because we aren't
+trying to auto complete the word(s) and current cell.
+
+If we mimicked ghost text. The experience might look as follows
+
+* User is editing cell N in the notebook
+* Foyle is continuously running in the background to generate suggested cells N+1,..., N+K
+* The suggested cells N+1,...,N+k are rendered in the notebook in a manner that makes it clear that they are suggestions
+  that haven't been accepted or rejected yet
+  * Following Ghost Text we could use a grayed out font and different border for the cells 
+* As the user edits cell N, Foyle updates and edits N+1,...,N+K to reflect changes to its suggestions
+* The decision to accept or reject the situation is triggered when the user decides to move onto cell N+1 
+  * If the user switches focus to the suggested N+1 cell that's considered accepting the suggestion
+  * If the user inserts a new cell that's considered a rejection
+* Each time Foyle generates a new completion it replaces any previous suggestions that haven't been accepted and inserts
+  the new suggested cells
+* When the notebook is persisted unapproved suggestions are pruned and not saved 
+
+Since every notebook cell is just a TextEditor I think we can use the 
+[TextEditorDecoration API](https://github.com/microsoft/vscode/blob/ea1445cc7016315d0f5728f8e8b12a45dc0a7286/src/vscode-dts/vscode.d.ts#L10521)
+to change how text is rendered in the suggested cells to indicate they are unaccepted suggestions.
+
+We can use [Notebook Cell Metadata](https://github.com/stateful/vscode-runme/blob/686f9f0904d65f061d2a270b8de3ea8a4ede9b0f/src/extension/ai/generate.ts#L95)
+to keep track of cells that are associated with a particular suggestion and need to be accepted or rejected. Since
+metadata is just a key value store we can add a boolean "unaccepted" to indicate a cell is still waiting on acceptance.
+
+## Completion Protocol
+
+Our existing[GenerateService](https://github.com/jlewi/foyle/blob/4e2db25378712e9b33d8257aae4f4a62d0edfa49/protos/foyle/v1alpha1/agent.proto#L27)
+is a unary RPC call. This isn't ideal for continually generating suggestions as a user updates a block.
+We can use the [connect protocol](https://connectrpc.com/docs/protocol/#streaming-rpcs) to define a full streaming RPC 
```

Comment:
Bi-/Uni-directional streams are part of the gPRC spec https://grpc.io/docs/what-is-grpc/core-concepts/#bidirectional-streaming-rpc which we're already using in `Runner#Execute` and other places.

Reply:
Yup. I think the advantage of connect is that if we use the connect server (which I've started to do in Foyle). Then it out of box gives you 3 protocols
* Connect
* grpc
* grpc-web

A big advantage here is that connect and grpc-web work in the browser. So we aren't introducing more blockers to being able to run the frontend in the browser (stateful/runme#616). 

I also think connect out of box supports JSON payloads which can make debugging/development easier.